Fix overly restrictive usage of authGSSServerInit

twosigmajab opened the pull request:
Drop hostname and service, to no longer needlessly limit to those passed at initialization time.

twosigmajab commented on setup.py:
```diff
@@ -16,7 +16,7 @@
 
 setup(name='WSGI-Kerberos',
-      version='0.2.0',
+      version='0.3.0',
```
As per semver.org,

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

Figured these changes merited this minor version bump.

twosigmajab commented:
@hawicz has reviewed and approved these changes. LGTY, @mkomitee?

mkomitee commented on wsgi_kerberos.py:
```diff
@@ -66,23 +66,15 @@ class KerberosAuthMiddleware(object):
     '''
 
-    def __init__(self, app, hostname=None, unauthorized=None, forbidden=None,
+    def __init__(self, app, unauthorized=None, forbidden=None,
```
I don’t think this is a 100% good change. I think the doc change to not push people to setting the hostname is good, and a change in behavior when it’s not provided (still None) is warranted to make the default to not restrict things, but dropping the option to restrict it isn’t something I’d support.

hawicz replied:
If you really want to allow people that level of low-level control, I think it'd be better to provide a service_name param instead of a hostname, to allow full customization of the service name to be used, and to discourage existing uses of the hostname param from being kept.

mkomitee replied:
And by service name, you mean allow them to override the `HTTP` part of the gss service name? This is for wsgi services, so I don’t see that as being all that beneficial, as wsgi pretty much implies http.

I also don’t think that or the existing proposed change are sufficient to warrant a change in API for users to deal with.

twosigmajab replied:
@mkomitee, I have reworked this changeset in response to your concerns. I no longer drop the `hostname` parameter from the constructor, allowing users who do pass it to still see a warning if no corresponding server principal is found, but preserve the change to pass `''` rather than the result of `socket.gethostname()` to the `kerberos.authGSSServerInit()` call, which is the crux of the fix for #3. Please let me know if this looks better. Thanks!

propertone reviewed (changes requested):
Few things:
* The API should pass the hostname to the `kerberos.authGSSServerInit` call if it was specified. Explicitly passing a hostname that's not in the Keytab should be an error.
* Please remove release notes/version changes (I'll add those separately when cutting a new release depending on what's actually included)
(FYI: I've taken over as primary project maintainer)

twosigmajab commented:
@propertone, I have removed the release notes and version changes in the latest revision.

However I am not sure about:

> The API should pass the hostname to the `kerberos.authGSSServerInit` call if it was specified.

Shouldn't the server's keytab be the single source of truth for valid service names? Wouldn't it be better to get WSGI-Kerberos out of the business of doing its own additional validation, instead allowing it to be centralized inside the GSS API?

Also, did you notice the change to accept an iterable of hostnames for the `hostname` argument, since services can have multiple keys? If the caller provided more than one hostname to `__init__`, it's ambiguous which one should be passed to `kerberos.authGSSServerInit` if WSGI-Kerberos is not getting out of the business of validating extra constraints imposed at initialization time, on top of what the GSS API is already doing at authentication time.

As for:

> Explicitly passing a hostname that's not in the Keytab should be an error.

If the client attempts to access the service via a name that is not in the keytab, it will still cause an error at authentication time, even with the change to pass `''` to `authGSSServerInit`.

Could you please take another look and let me know what you think?

Thanks, and nice to see this project is now being maintained again!

propertone commented:
In most cases, the hostname would just be left out, which is the recommended approach.

However the option to specify the server principal is consistent with many other applications that implement Kerberos, such as [mod_auth_gssapi](https://github.com/gssapi/mod_auth_gssapi#gssapiacceptorname) and [nginx spnego module](https://github.com/stnoonan/spnego-http-auth-nginx-module). 

I don't really see the need to support multiple hostnames. The use case for explicitly setting the hostname is pretty rare and more for preserving existing behavior (of failing if the specified hostname is not in the keytab file).

twosigmajab committed:
Don't needlessly limit to service resolved at init time.

Previously, when the optional `hostname` argument was not passed to
__init__, it would fall back to `socket.gethostname()`.

__init__ would then populate `self.service` based on the resulting
`hostname` value, and this same `self.service` value would then always
be passed to `authGSSServerInit`.

However, pre-specifying a service when calling `authGSSServerInit`
needlessly constrains what service names can work and breaks any clients
that attempt to access the service by another name, even if a key for
that service name exists in the keytab by the time authentication
happens.

Now, when `hostname` is not passed to to __init__, empty string is passed
to `authGSSServerInit` to allow clients accessing the service by another
name (for which a key exists in the keytab by the time it is called) to
do so successfully.

Fixes #3.

twosigmajab commented:
@propertone, I've removed the ability to pass multiple hostnames, and restored the behavior where any hostname provided is used for the `authGSSServerInit` calls. Another look?

propertone commented on wsgi_kerberos.py:
```diff
@@ -81,11 +83,11 @@ class KerberosAuthMiddleware(object):
-
+        if hostname:
+            self._check_hostname(i)
+        self.service = hostname and 'HTTP@%s' % hostname
```
Bug: `i => hostname`. Also please rewrite as if/else statement (`and` syntax is unnecessarily confusing)

twosigmajab replied:
Done in the latest revision.

propertone commented on wsgi_kerberos.py:
```diff
@@ -66,11 +66,13 @@ class KerberosAuthMiddleware(object):
-
     :param app: WSGI Application
-    :param hostname: Hostname for kerberos name canonicalization
+    :param hostname: Optional hostname to check whether a corresponding server
+        principal has been configured at init time (if not, a warning will be
+        logged), and to use when calling authGSSServerInit at authentication
+        time. If not provided, empty string will be passed to authGSSServerInit
+        so that clients can access the service by any name for which a key
+        exists in the keytab.
```
A technically correct description, but maybe confusing to people unfamiliar with kerberos internals. How about something simpler like this:
```
Force the server to only accept requests for the specified hostname. If not specified, clients can access the service by any name in the keytab.
```

twosigmajab replied:
Good call! Done in the latest revision.

twosigmajab commented on wsgi_kerberos.py:
```diff
@@ -6,7 +6,6 @@
 import errno
 import kerberos
 import logging
-import os
```
(noticed this was unused while I was working on this, so I removed it while I was in here)

propertone reviewed (approved):
Looks good!
